src/LVMIndex.cpp

```diff
@@ -0,0 +1,479 @@
+        bool less(const TupleRef& x, const TupleRef& y) const {
+            return operator()(x, y) < 0;
+        }
+
+        /* equal comparison */
+        bool equal(const TupleRef& x, const TupleRef& y) const {
+            for (auto& i : order) {
+                if (x[i] != y[i]) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    };
+
+    class Source : public Stream::Source {
+        // the begin and end of the stream
+        using iter = btree_multiset<TupleRef, comparator, std::allocator<TupleRef>, 512>::iterator;
+        iter cur;
+        iter end;
+
+        // an internal buffer for re-ordered elements
+        std::array<TupleRef, Stream::BUFFER_SIZE> buffer;
+
+    public:
+        Source(iter begin, iter end) : cur(begin), end(end) {}
+
+        int load(TupleRef* out, int max) override {
+            int c = 0;
+            while (cur != end && c < max) {
+                buffer[c] = *cur;
+                out[c] = buffer[c];
+                ++cur;
+                ++c;
+            }
+            return c;
+        }
+
+        std::unique_ptr<Stream::Source> clone() override {
+            auto source = std::make_unique<Source>(cur, end);
+            source->buffer = this->buffer;
+            return source;
+        }
+    };
+
+    /* btree for storing tuple pointers with a given lexicographical order */
+    using index_set = btree_multiset<TupleRef, comparator, std::allocator<TupleRef>, 512>;
+
+    IndirectIndex(std::vector<int> order)
+            : theOrder(std::move(order)), set(comparator(theOrder), comparator(theOrder)),
+              Arity(order.size()) {}
+
+    size_t arity() const override {
+        return Arity;
+    };
+
+    bool empty() const override {
+        return set.empty();
+    }
+
+    size_t size() const override {
+        return set.size();
+    }
+
+    bool insert(const TupleRef& tuple) override {
+        return set.insert(tuple, operation_hints);
+    }
+
+    void insert(const LVMIndex& src) override {
+        for (const auto& cur : src.scan()) {
+            insert(cur);
+        }
+    }
+
+    bool contains(const TupleRef& tuple) const override {
+        return set.contains(tuple);
+    }
+
+    Stream scan() const override {
+        return std::make_unique<Source>(set.begin(), set.end());
+    }
+
+    Stream range(const TupleRef& low, const TupleRef& high) const override {
+        return std::make_unique<Source>(set.lower_bound(low), set.upper_bound(high));
+    }
+
+    void clear() override {
+        set.clear();
+    }
+
+private:
+    /** retain the index order used to construct an object of this class */
+    const std::vector<int> theOrder;
+
+    /** set storing tuple pointers of table */
+    index_set set;
+
+    /** Operation hints */
+    index_set::btree_operation_hints<1> operation_hints;
+
+    /** Arity as the relation arity, not necessary the order size in indirect index */
+    size_t Arity;
```

Comment:
Is this intended to be capitalised? It looks like a standard variable, thus should be lower cased.